### ISSUE TITLE
Removing albumentations from model

### ISSUE THREAD

**Issue by stavMarz** [ultralytics/ultralytics#4254]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello , when I train the model on my custom dataset and change the parameters accordingly, I see in the execution that there is albumentations on my dataset . I have thermal images and I dont want any blur on my dataset. 

![ablu](https://github.com/ultralytics/ultralytics/assets/100599243/e2392392-5d32-4cc7-835e-532cecd648e0)
This is my code: 

`from ultralytics import YOLO`
`model = YOLO('yolov8x.yaml')  # build a new model from YAML`
`model = YOLO('yolov8x.pt')  # load a pretrained model (recommended for training)`
`model = YOLO('yolov8x.yaml').load('yolov8x.pt')  # build from YAML and transfer weights`
`model.train(data='/content/dataset/data.yaml', epochs=200, imgsz=640, save_period = 50, project ="A", seed = 123,dropout = 0.01, val = True, hsv_h = 0, hsv_s = 0, hsv_v = 0, degrees = 20, translate = 0, scale = 0.5 , fliplr= 0, mosaic= 0 )`

Is it possible to cancel the cancel the albumentations applied on my dataset? 
If so, I would love to know how or to get any other relevant information regarding this issue!

Thanks a lot, 
Stav

**Comment by stavMarz** [comment]
Hii @glenn-jocher thank you so much for the quick response!
In the configuration file YOLOV8.yaml there isn't any albumentations configuration. (I'm using the default setting, I didn't change anything)
So I am not sure if and how I can remove them...
Also, is it possible that the logger prints the types of albumentations but doesn't necessarily use them? 

Thank you again for the help, much appreciated!

**Comment by glenn-jocher** [comment]
Hello @stavMarz,

Thank you for your response and clarification.

In YOLOv8, the default setting doesn't include specific albumentations in the configuration file. The image augmentations performed are integrated into the dataset loader and are not part of the configuration file. 

What's likely happening here is - some form of data augmentation is being applied by default when you run your train method, even if it's not explicitly stated in the yaml file. These augmentations are designed to improve model performance by artificially expanding the training data. 

When it comes to the logger, yes, it can log messages about the types of possible albumentations, but it doesn't necessarily mean all of them are being used.

In case you do want to turn off specific augmentations during training, you would need to modify the source code where the dataset loader applies these augmentations.

I hope this clears up the confusion. Feel free to ask if you have any more questions.

Best regards,
Glenn

**Comment by andresinsitu** [comment]
@glenn-jocher wheres the logger file located?

**Comment by glenn-jocher** [comment]
@andresinsitu hello,

Thank you for your suggestion! Indeed, managing the augmentation pipeline from the yaml file or `model.train()` parameters would bring a higher level of flexibility to the process and ease of use.

Your workaround of uninstalling albumentations is quite clever as well. As you correctly noticed, the source code checks whether this library is installed before attempting to use it.

We will take your suggestion into consideration for future updates of YOLOv8.

Thank you for your feedback and contribution to improving the YOLOv8 framework!

Best regards,
[Assistant]

**Comment by khm159** [comment]
I also require this function. 
I add two lines at original codebase... 

In Ultralytics, there are no option to change the algumentation probability and always p=1.0.
And also if you add the implement the custom Trainer, you can not add new argument not in default.yaml... 

Therefore i change the default.yaml in internel Ulytalytics it self. 

https://github.com/ultralytics/ultralytics/pull/4604
  
You can turn-off basic Albumentation for changing the albumentation probablility to 0 in your overridding yaml...

**Comment by aymuos15** [comment]
Hi, for the time being, is there any way to mention how to choose which augmentations one wants to use?
Would be very helpful if I could be directed to that particular path.

**Comment by khm159** [comment]
> Hi, for the time being, is there any way to mention how to choose which augmentations one wants to use? Would be very helpful if I could be directed to that particular path.

Hi

You can directly change your Ultralytics library

In youlv8, the full augmentation pipeline is like this (as i understood)

1. The data augmentation are automatically builed while building [YOLODataset](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/dataset.py#L135)

2. The YOLODataset override build_transforms() function of [BaseDataset](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/base.py#L23).

3. In [build_transforms()](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/base.py#L23) 

There are series of augmentation is automatically builded

```
if self.augment:
    hyp.mosaic = hyp.mosaic if self.augment and not self.rect else 0.0
    hyp.mixup = hyp.mixup if self.augment and not self.rect else 0.0
    transforms = v8_transforms(self, self.imgsz, hyp)
else:
    transforms = Compose([LetterBox(new_shape=(self.imgsz, self.imgsz), scaleup=False)])
    transforms.append(
        Format(
            bbox_format='xywh',
            normalize=True,
            return_mask=self.use_segments,
            return_keypoint=self.use_keypoints,
            batch_idx=True,
            mask_ratio=hyp.mask_ratio,
            mask_overlap=hyp.overlap_mask))
        return transforms
```

As you can see, if you turn on the data augmentation option, then the v8_transforms are builted. And the problem is here. 

4. [v8_transforms](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/augment.py#L767)

```
def v8_transforms(dataset, imgsz, hyp, stretch=False):
    """Convert images to a size suitable for YOLOv8 training."""
    pre_transform = Compose([
        Mosaic(dataset, imgsz=imgsz, p=hyp.mosaic),
        CopyPaste(p=hyp.copy_paste),
        RandomPerspective(
            degrees=hyp.degrees,
            translate=hyp.translate,
            scale=hyp.scale,
            shear=hyp.shear,
            perspective=hyp.perspective,
            pre_transform=None if stretch else LetterBox(new_shape=(imgsz, imgsz)),
        )])
    flip_idx = dataset.data.get('flip_idx', [])  # for keypoints augmentation
    if dataset.use_keypoints:
        kpt_shape = dataset.data.get('kpt_shape', None)
        if len(flip_idx) == 0 and hyp.fliplr > 0.0:
            hyp.fliplr = 0.0
            LOGGER.warning("WARNING ⚠️ No 'flip_idx' array defined in data.yaml, setting augmentation 'fliplr=0.0'")
        elif flip_idx and (len(flip_idx) != kpt_shape[0]):
            raise ValueError(f'data.yaml flip_idx={flip_idx} length must be equal to kpt_shape[0]={kpt_shape[0]}')

    return Compose([
        pre_transform,
        MixUp(dataset, pre_transform=pre_transform, p=hyp.mixup),
        Albumentations(p=1.0),
        RandomHSV(hgain=hyp.hsv_h, sgain=hyp.hsv_s, vgain=hyp.hsv_v),
        RandomFlip(direction='vertical', p=hyp.flipud),
        RandomFlip(direction='horizontal', p=hyp.fliplr, flip_idx=flip_idx)])  # transforms

```

As you can see, the Albumentations(**p=1.0**)

If you turn on the data augmentation, then Albumentation(blur, gray scale etc) are **automatically adopted, and can not disable it.** 

Therefore If you want to turn off Albumentation, then **you should revise Library ITSELF**

For example, If you want to turn off blur, then ... You should change code of [Albumentation wrapper classes](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/data/augment.py#L646) 

Because in the Albumentation wrapper, **the all probabilities are fixed value, and can not accesse the overridding yaml. **

```
            T = [
                A.Blur(p=0.01),
                A.MedianBlur(p=0.01),
                A.ToGray(p=0.01),
                A.CLAHE(p=0.01),
                A.RandomBrightnessContrast(p=0.0),
                A.RandomGamma(p=0.0),
                A.ImageCompression(quality_lower=75, p=0.0)]  # transforms
            self.transform = A.Compose(T, bbox_params=A.BboxParams(format='yolo', label_fields=['class_labels']))
```

**I think it is trickey and should be able to access by overriding yaml file...**
 

Or you can try like this... **(Because the yaml argument not in default.yaml is blocked)**

https://github.com/ultralytics/ultralytics/pull/4604/commits/83a5dd33f6432c1b5eb89345a1bf7892e0bc38dd

**Comment by aymuos15** [comment]
Please correct me if i'm wrong, but assuming I want to set: `A.CLAHE(p=0.01),` to 0, then i just have to clone the repo locally and make the change in that exact file path?

And thank you so much for such a detailed explanation. This is super helpful.

**Comment by glenn-jocher** [comment]
@aymuos15 yes, you are correct. If you want to set `A.CLAHE(p=0.01)` to 0, you would have to clone the repository locally and make the change in the exact file and line where this augmentation is defined.

As for your suggestion about the disclaimer, that's a great point. While these augmentations are designed to improve model performance on a wide range of datasets, they may not always be beneficial for very specific or niche tasks. We will take your feedback into consideration and look at ways to provide clearer communication about the built-in augmentations in the repository's documentation.

We always strive to make the repository as user-friendly and transparent as possible, so we appreciate your valuable input.

Thank you for bringing this to our attention and for your continued support of YOLOv8!

**Comment by khm159** [comment]
Thank you for your consideration. yolo-v8 is a tremendous help and thank you. @glenn-jocher 

I'm glad it was helpful. @aymuos15

**Comment by glenn-jocher** [comment]
Hello @khm159,

We're delighted to hear that YOLOv8 has been a significant asset for your research. User feedback is important to us as it helps us continuously improve and develop. Feel free to share any enhancement suggestions or issues you might encounter. We're committed to providing a robust and efficient tool for the community.

Best Regards,
[Assistant]

**Comment by aymuos15** [comment]
@glenn-jocher Yes, I do realize that the augmentations are shown anyways, but printing it in a slightly more 'in your face' way might get it more attention? Not too sure haha.

Either way, for me too, this repo and YOLOv8 has been extremely regarding. Thank you for the quick replies as well.

**Comment by glenn-jocher** [comment]
@aymuos15 hello,

Thank you for your suggestion. You make a great point about making augmentations more apparent and we will certainly take that into consideration as we aim to make YOLOv8 as user-friendly as possible. Our goal is to provide clear and helpful information that assists users in effectively using the repository.

We are also glad to hear that YOLOv8 has been beneficial for you. Your appreciation and feedback mean a lot to us, they help us constantly improve and adapt to the needs of our users. Feel free to reach out if you have any more suggestions or questions.

Thank you for your continuous support!

Best regards,
[Assistant]

**Comment by Mehreganbeiki** [comment]
Hi,

In a related matter, it was previously mentioned that to disable augmentation, we should set the `augment` parameter to `False` in the default YAML file. However, I've noticed that the `augment` setting is located within the prediction block of the YAML file. This has raised a question for me: Is there a method to prevent the model from applying augmentation during the training process as well? My concern arises from the presence of various parameters such as `hsv_h`, `hsv_s`, `shear`, `perspective`, `flipud`, `fliplr`, and `mosaic` in the hyperparameters section, as well as commands related to horizontal flipping and color jitter transformations in the command line under Albumentations. This made me confused about whether setting the `augment` parameter in the default YAML file is the correct approach to disable augmentation.

**Comment by glenn-jocher** [comment]
Yes you can set any augmentation you want to zero to disable it, or set them all to zero to disable all augmentations.
